k8s_api_client/README.md

```diff
@@ -0,0 +1,157 @@
+              name: opa-kubernetes-api-client
+              key: token
+```
+
+We may now retrieve the service account token from inside our rego policies by calling `opa.runtime().env.KUBERNETES_API_TOKEN`.
+
+### RBAC
+
+In order to actually read any data, you'll need to allow your service account to do so. For this you'll need to create an RBAC role and role binding. In the example below we'll create a role that has cluster level read access to some common resource types, but you'll naturally want to tweak this for your needs:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: opa-kubernetes-api-client
+rules:
+- apiGroups: [""]
+  resources: ["services", "pods", "configmaps"]
+  verbs: ["get", "list"]
+- apiGroups: ["apps"]
+  resources: ["daemonsets", "deployments", "statefulsets", "replicasets"]
+  verbs: ["get", "list"]
+```
+
+Let's bind the role to our service account:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: opa-kubernetes-api-client-read
+subjects:
+- kind: ServiceAccount
+  name: opa-kubernetes-api-client
+  namespace: default
+roleRef:
+  kind: ClusterRole
+  name: opa-kubernetes-api-client
+  apiGroup: rbac.authorization.k8s.io
+```
+
+All done! You should now be able to call the functions included in this library to query the kubernetes API for resources directly.
+
+## Available functions
+
+```rego
+# Query for given resource/name in provided namespace
+# Example: query_ns("deployments", "my-app", "default")
+query_name_ns(resource, name, namespace)
+
+# Query for given resource type using label selectors
+# https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api
+# Example: query_label_selector_ns("deployments", {"app": "opa-kubernetes-api-client"}, "default")
+query_label_selector_ns(resource, selector, namespace)
+
+# Query for all resources of type resource in all namespaces
+# Example: query_all("deployments")
+query_all(resource)
+```
+
+## Local development and testing
+
+Run the `setup.sh` script to create a local kubernetes cluster using [kind](https://kind.sigs.k8s.io/). This will deploy OPA and an ingress controller to let you access it as if running normally (i.e. on `localhost:8181`). You can then run queries using the [OPA query API](https://www.openpolicyagent.org/docs/latest/rest-api/#query-api) in order to test the functions from the policy.
+
+`POST http://localhost:8181/v1/query`
+```json
+{
+    "query": "x := data.kubernetes.api.client.query_name_ns(\"deployments\",\"opa-kubernetes-api-client\", \"default\").body"
+}
+```
+
+`POST http://localhost:8181/v1/query`
+```json
+{
+    "query": "x := data.kubernetes.api.client.query_all(\"deployments\").body"
+}
+```
+
+FIXME
+
+`POST http://localhost:8181/v1/query`
+```json
+{
+    "query": "x := data.kubernetes.api_client.query_label_selector_ns(\"deployments\", {\"app\":\"opa-kubernetes-client\"}, \"default\")"
+}
+```
```

Comment:
🤔 Would you think it worthwhile to turn these into a script, or even a test rego file, so that these instructions can be verified against the running kind cluster?

Reply:
That's a good idea 👍 I'll add that.

Reply:
Done :)

Reply:
🚀 Thanks!